### PR TITLE
[C10D] Clarify warning for concurrent PG usage

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131895

Addresses a common misconception about safety of using multiple NCCL
process groups from PyTorch.

Notably, it IS safe to use multiple process groups, so long as
communication operations from different groups are not allowed to
overlap.  (Overlap of communication operations from one group with
compute operations IS ok).

TODO: after getting feedback on the text, update other copies of the warning on other APIs

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -4515,12 +4515,20 @@ def new_group(
...
-        their application to ensure only one process group is used at a time.
-        This means collectives from one process group should have completed
-        execution on the device (not just enqueued since CUDA execution is
-        async) before collectives from another process group are enqueued.
+        When using multiple process groups with the ``NCCL`` backend, the user
+        must ensure a globally consistent execution order of collectives across
+        ranks.
+
+        If multiple threads within a process issue collectives, explicit
+        synchronization is necessary to ensure consistent ordering.
```
Comment (KaimingOuyang) [inline torch/distributed/distributed_c10d.py:4523]: How about:
If multiple threads within a process issue collectives to different process groups, explicit synchronization between threads is necessary to ensure consistent ordering.

Comment (wconstab) [reply torch/distributed/distributed_c10d.py:4523]: ok- so its actually fine for multiple threads in the process to issue collectives to one PG without doing synchronization? i thought that would also be wrong

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -4515,12 +4515,20 @@ def new_group(
...
+
+        When using async variants of torch.distributed communication APIs,
+        a work object is returned and the communication kernel is
+        enqueued on a separate CUDA stream, allowing overlap of communication
+        and computation. Once one or more async ops have been issued on one process
+        group, they must be synchronized by calling `work.wait()` before using
```
Comment (fduwjj) [inline torch/distributed/distributed_c10d.py:4529]: Do we also want to mention that this type of synchronization is not a sync between CPU and GPU? (unless CUDA_LAUNCH_BLOCK is set)

**Review by fduwjj — APPROVED** [review]

File: torch/distributed/distributed_c10d.py (hunk 2 above)
Comment (wconstab) [reply torch/distributed/distributed_c10d.py:4529]: ill tweak the wording here to reflect that

**Pull request closed without merging**